File: mobile/src/main/java/com/gitlab/PCU/PCU/ServerSettingsActivity.java
package com.gitlab.PCU.PCU;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.ArrayMap;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.gitlab.PCU.PCU.helper.Defaults;
import com.gitlab.PCU.PCU.helper.ServerCfg;
import com.gitlab.PCU.PCU.helper.ServerSettingsStore;

import java.util.ArrayList;
import java.util.Map;

import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;
import static com.gitlab.PCU.PCU.helper.Defaults.ResultCode.RESULT_DELETE;

public class ServerSettingsActivity extends AppCompatActivity {

    private ArrayList<View> childs = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_server_settings);
        createViews();
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        try {
            //noinspection ConstantConditions
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        } catch (NullPointerException e) {
            System.out.print("NPS @getSupportActionBar().setDisplayHomeAsUpEnabled(true)");
        }
    }

    private void createViews() {
        for (View child : childs) {
            ((ViewGroup) child.getParent()).removeView(child);
        }

        childs = new ArrayList<>();

        ArrayMap<String, String[]> list = ServerCfg.list(getSharedPreferences("servers", 0));

        Context context = getApplicationContext();

        LinearLayout linearLayout = (LinearLayout) findViewById(R.id.content_server_settings_ll);

        for (Map.Entry<String, String[]> listEntry : list.entrySet()) {
            TextView name = new TextView(context);
            name.setText(listEntry.getValue()[0]);
            TextView info = new TextView(context);
            info.setText(listEntry.getValue()[1]);
            String id = listEntry.getKey();
            addServerView(name, info, linearLayout, id);
        }
    }

    private void addServerView(TextView name, TextView info, LinearLayout linearLayout, String serverId) {
        if (linearLayout.getChildCount() > 0) {
            View spacer = new View(getApplicationContext());
            spacer.setLayoutParams(new ViewGroup.LayoutParams(MATCH_PARENT, 2));
            spacer.setBackgroundColor(getResources().getColor(R.color.colorLightGray, getTheme()));
            linearLayout.addView(spacer);
            childs.add(spacer);
        }

        name.setTextSize(18);
        info.setTextSize(14);

        info.setTextColor(getResources().getColor(R.color.colorLightGray, getTheme()));

        EditOCL editOCL = new EditOCL(serverId);

        name.setOnClickListener(editOCL);
        info.setOnClickListener(editOCL);

        linearLayout.addView(name);
        childs.add(name);
        linearLayout.addView(info);
        childs.add(info);
    }

    public void onFab(View v) {
        Context context = getApplicationContext();
        ArrayMap<String, String[]> arrayMap = ServerCfg.list(getSharedPreferences("servers", 0));
        int i = arrayMap.size();
        while (arrayMap.containsKey("Server" + Integer.toString(i))) {
            i++;
        }
        String id = "Server" + Integer.toString(i);
        ServerSettingsStore serverSettingsStore = new Defaults.ServerVariable(context).SERVER_SETTINGS_STORE;
        ServerCfg.write(getSharedPreferences("servers", 0), id, serverSettingsStore);
        TextView name = new TextView(context);
        name.setText(serverSettingsStore.getName());
        TextView info = new TextView(context);
        info.setText(serverSettingsStore.getDesc());
        addServerView(name, info, (LinearLayout) findViewById(R.id.content_server_settings_ll), id);
    }

    private void startEdit(ServerSettingsStore serverSettingsStore, String serverId) {
        Intent intent = new Intent(getApplicationContext(), SingleServerSettings.class);
        intent.putExtra("in", serverSettingsStore);
        intent.putExtra("id", serverId);
        startActivityForResult(intent, Defaults.RequestCode.EDIT_SERVER);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode) {
            case Defaults.RequestCode.EDIT_SERVER:
                if (resultCode == RESULT_OK && data != null) {
                    ServerSettingsStore serverSettingsStore = data.getParcelableExtra("out");
                    String id = data.getStringExtra("id");
                    ServerCfg.write(getSharedPreferences("servers", 0), id, serverSettingsStore);
                    createViews();
                } else if (resultCode == RESULT_DELETE && data != null) {
                    String id = data.getStringExtra("id");
                    ServerCfg.delete(getSharedPreferences("servers", 0), id);
                    createViews();
                }
                break;
        }
    }

    private class EditOCL implements View.OnClickListener {

        private final String id;

        EditOCL(String id) {
            this.id = id;
        }

        /**
         * Called when a view has been clicked.
         *
         * @param v The view that was clicked.
         */
        @Override
        public void onClick(View v) {
            ServerSettingsStore serverSettingsStore = ServerCfg.read(getSharedPreferences("servers", 0), new Defaults.ServerVariable(getApplicationContext()), id);
            startEdit(serverSettingsStore, id);
        }
    }
}

File: mobile/src/main/java/com/gitlab/PCU/PCU/ServerSettingsActivity.kt
package com.gitlab.PCU.PCU

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.LayoutParams.MATCH_PARENT
import android.widget.LinearLayout
import android.widget.TextView
import com.gitlab.PCU.PCU.R.layout.activity_server_settings
import com.gitlab.PCU.PCU.helper.Defaults
import com.gitlab.PCU.PCU.helper.Defaults.ResultCode.RESULT_DELETE
import com.gitlab.PCU.PCU.helper.ServerCfg
import com.gitlab.PCU.PCU.helper.ServerSettingsStore
import kotlinx.android.synthetic.main.activity_main_screen.*
import java.util.*


class ServerSettingsActivity : AppCompatActivity() {

    private var childs = ArrayList<View>()
    private val adapter: RecyclerView.Adapter<*>? = null
    private val layoutManager: RecyclerView.LayoutManager? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(activity_server_settings)
        try {
            createViews()
        } catch (ignored: NullPointerException) {
        }

        setSupportActionBar(toolbar)
        try {

            supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        } catch (e: NullPointerException) {
            print("NPS @ getSupportActionBar().setDisplayHomeAsUpEnabled(true)")
        }

    }

    private fun createViews() {
        for (child in childs) {
            (child.parent as ViewGroup).removeView(child)
        }

        childs = ArrayList<View>()

        val list = ServerCfg.list(getSharedPreferences("servers", 0))

        val context = applicationContext

        for ((id, value) in list) {
            val name = TextView(context)
            name.text = value[0]
            val info = TextView(context)
            info.text = value[1]
            addServerView(name, info, content_server_settings_ll, id)
        }
    }

    private fun addServerView(name: TextView, info: TextView, linearLayout: LinearLayout, serverId: String) {
        if (linearLayout.childCount > 0) {
            val spacer = View(applicationContext)
            spacer.layoutParams = ViewGroup.LayoutParams(MATCH_PARENT, 2)
            spacer.setBackgroundColor(resources.getColor(R.color.colorLightGray, theme))
            linearLayout.addView(spacer)
            childs.add(spacer)
        }

        name.textSize = 18f
        info.textSize = 14f

        info.setTextColor(resources.getColor(R.color.colorLightGray, theme))

        val editOCL = EditOCL(serverId)

        name.setOnClickListener(editOCL)
        info.setOnClickListener(editOCL)

        linearLayout.addView(name)
        childs.add(name)
        linearLayout.addView(info)
        childs.add(info)
    }

    fun onFab(v: View) {
        val context = applicationContext
        val arrayMap = ServerCfg.list(getSharedPreferences("servers", 0))
        var i = arrayMap.size
        while (arrayMap.containsKey("Server" + Integer.toString(i))) {
            i++
        }
        val id = "Server" + Integer.toString(i)
        val serverSettingsStore = Defaults.ServerVariable(context).SERVER_SETTINGS_STORE
        ServerCfg.write(getSharedPreferences("servers", 0), id, serverSettingsStore)
        val name = TextView(context)
        name.text = serverSettingsStore.getName()
        val info = TextView(context)
        info.text = serverSettingsStore.getDesc()
        addServerView(name, info, findViewById(R.id.content_server_settings_ll) as LinearLayout, id)
    }

    private fun startEdit(serverSettingsStore: ServerSettingsStore, serverId: String) {
        val intent = Intent(applicationContext, SingleServerSettings::class.java)
        intent.putExtra("in", serverSettingsStore)
        intent.putExtra("id", serverId)
        startActivityForResult(intent, Defaults.RequestCode.EDIT_SERVER)
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            Defaults.RequestCode.EDIT_SERVER -> if (resultCode == Activity.RESULT_OK && data != null) {
                val serverSettingsStore = data.getParcelableExtra<ServerSettingsStore>("out")
                val id = data.getStringExtra("id")
                ServerCfg.write(getSharedPreferences("servers", 0), id, serverSettingsStore)
                createViews()
            } else if (resultCode == RESULT_DELETE && data != null) {
                val id = data.getStringExtra("id")
                ServerCfg.delete(getSharedPreferences("servers", 0), id)
                createViews()
            }
        }
    }

    private inner class EditOCL internal constructor(private val id: String) : View.OnClickListener {

        /**
         * Called when a view has been clicked.

         * @param v The view that was clicked.
         */
        override fun onClick(v: View) {
            val serverSettingsStore = ServerCfg.read(getSharedPreferences("servers", 0), Defaults.ServerVariable(applicationContext), id)
            startEdit(serverSettingsStore, id)
        }
    }
}
